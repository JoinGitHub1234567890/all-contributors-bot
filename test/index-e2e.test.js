const nock = require('nock')
const { Probot } = require('probot')

const allContributorsApp = require('../src')
const { rejectionOf } = require('./testUtils')

const issue_commentCreatedPayload = require('./fixtures/issue_comment.created.json')
const reposGetContentsAllContributorsRCdata = require('./fixtures/repos.getContents.all-contributorsrc.json')
const usersGetByUsernameJakeBolamdata = require('./fixtures/users.getByUsername.jakebolam.json')
const reposGetContentsREADMEMDdata = require('./fixtures/repos.getContents.README.md.json')

describe('All Contributors app - End to end', () => {
    let probot

    beforeEach(() => {
        probot = new Probot({})
        const app = probot.load(allContributorsApp)

        // just return a test token
        app.app = () => 'test'
    })

    test('Happy path, add correct new contributor', async () => {
        nock('https://api.github.com')
            .post('/app/installations/11111/access_tokens')
            .reply(200, { token: 'test' })

        nock('https://api.github.com')
            .get(
                '/repos/all-contributors/all-contributors-bot/contents/.all-contributorsrc',
            )
            .reply(200, reposGetContentsAllContributorsRCdata)

        nock('https://api.github.com')
            .get('/users/jakebolam')
            .reply(200, usersGetByUsernameJakeBolamdata)

        nock('https://api.github.com')
            .get(
                '/repos/all-contributors/all-contributors-bot/contents/README.md',
            )
            .reply(200, reposGetContentsREADMEMDdata)

        nock('https://api.github.com')
            .post(
                '/repos/all-contributors/all-contributors-bot/issues/1/comments',
                body => {
                    expect(body).toMatchSnapshot()
                    return true
                },
            )
            .reply(200)

        await probot.receive({
            name: 'issue_comment',
            payload: issue_commentCreatedPayload,
        })
    })

    test('Fail path, no readme file (configuration error)', async () => {
        nock('https://api.github.com')
            .post('/app/installations/11111/access_tokens')
            .reply(200, { token: 'test' })

        nock('https://api.github.com')
            .get(
                '/repos/all-contributors/all-contributors-bot/contents/.all-contributorsrc',
            )
            .reply(200, reposGetContentsAllContributorsRCdata)

        nock('https://api.github.com')
            .get('/users/jakebolam')
            .reply(200, usersGetByUsernameJakeBolamdata)

        nock('https://api.github.com')
            .get(
                '/repos/all-contributors/all-contributors-bot/contents/README.md',
            )
            .reply(404)

        nock('https://api.github.com')
            .post(
                '/repos/all-contributors/all-contributors-bot/issues/1/comments',
                body => {
                    expect(body).toMatchSnapshot()
                    return true
                },
            )
            .reply(200)

        await probot.receive({
            name: 'issue_comment',
            payload: issue_commentCreatedPayload,
        })
    })

    test('Fail path, no such user', async () => {
        nock('https://api.github.com')
            .post('/app/installations/11111/access_tokens')
            .reply(200, { token: 'test' })

        nock('https://api.github.com')
            .get(
                '/repos/all-contributors/all-contributors-bot/contents/.all-contributorsrc',
            )
            .reply(200, reposGetContentsAllContributorsRCdata)

        nock('https://api.github.com')
            .get('/users/jakebolam')
            .reply(404)

        nock('https://api.github.com')
            .post(
                '/repos/all-contributors/all-contributors-bot/issues/1/comments',
                body => {
                    expect(body).toMatchSnapshot()
                    return true
                },
            )
            .reply(200)

        await probot.receive({
            name: 'issue_comment',
            payload: issue_commentCreatedPayload,
        })
    })

    test('Fail path, no allcontributors file (repo needs setup)', async () => {
        nock('https://api.github.com')
            .post('/app/installations/11111/access_tokens')
            .reply(200, { token: 'test' })

        nock('https://api.github.com')
            .get(
                '/repos/all-contributors/all-contributors-bot/contents/.all-contributorsrc',
            )
            .reply(404)

        nock('https://api.github.com')
            .post(
                '/repos/all-contributors/all-contributors-bot/issues/1/comments',
                body => {
                    expect(body).toMatchSnapshot()
                    return true
                },
            )
            .reply(200)

        await probot.receive({
            name: 'issue_comment',
            payload: issue_commentCreatedPayload,
        })
    })

    test('Fail path, Unknown error (e.g. Network is dead, service down etc, our code is bad) crashes and sends error message', async () => {
        nock('https://api.github.com')
            .post('/app/installations/11111/access_tokens')
            .reply(200, { token: 'test' })

        nock('https://api.github.com')
            .get(
                '/repos/all-contributors/all-contributors-bot/contents/.all-contributorsrc',
            )
            .reply(500)

        nock('https://api.github.com')
            .post(
                '/repos/all-contributors/all-contributors-bot/issues/1/comments',
                body => {
                    expect(body).toMatchSnapshot()
                    return true
                },
            )
            .reply(200)

        const error = await rejectionOf(
            probot.receive({
                name: 'issue_comment',
                payload: issue_commentCreatedPayload,
            }),
        )
        expect(error instanceof Error).toBeTruthy()
        expect(error).toMatchSnapshot()
    })
})
