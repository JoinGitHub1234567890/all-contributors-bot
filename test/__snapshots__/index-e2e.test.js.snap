// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`All Contributors app - End to end Fail path, Unknown error (e.g. Network is dead, service down etc, our code is bad) crashes and sends error message 1`] = `
Object {
  "body": "@jakebolam 

We had trouble processing your request

Error: 500",
}
`;

exports[`All Contributors app - End to end Fail path, Unknown error (e.g. Network is dead, service down etc, our code is bad) crashes and sends error message 2`] = `[HttpError]`;

exports[`All Contributors app - End to end Fail path, no allcontributors file (repo needs setup) 1`] = `
Object {
  "body": "@jakebolam 

This project is not yet setup for [all-contributors](https://github.com/all-contributors/all-contributors).

    You will need to first setup [all-contributors-bot](https://github.com/all-contributors/all-contributors-bot) using the [all-contributors-cli](https://github.com/all-contributors/all-contributors-cli) tool.",
}
`;

exports[`All Contributors app - End to end Fail path, no readme file (configuration error) 1`] = `
Object {
  "body": "@jakebolam 

File README.md was not found in the repository (undefined).",
}
`;

exports[`All Contributors app - End to end Fail path, no such user 1`] = `
Object {
  "body": "@jakebolam 

Could not find the user jakebolam on github.",
}
`;

exports[`All Contributors app - End to end Happy path, add correct new contributor 1`] = `
Object {
  "body": "@jakebolam 

I've put up [a pull request](#1) to add jakebolam! :tada:",
}
`;
